README.md
# express-json-data

[![npm version](https://badge.fury.io/js/%40kjots%2Fexpress-json-data.svg)](https://www.npmjs.com/package/@kjots/express-json-data)

> A JSON data object Express router with PATCH support.

## Installation

```shell
npm install --save @kjots/express-json-data
```

## Usage

### As an ES6 module

```js
import express from 'express';
import expressJsonData from '@kjots/express-json-data';

let data = {};

let app = express()
    .use('/data', expressJsonData({ data }));
```

### As an ES5 module

```js
var express = require('express');
var expressJsonData = require('@kjots/express-json-data').default;

var data = {};

var app = express()
    .use('/data', expressJsonData({ data: data }));
```

## API

### expressJsonData(options)

Create an Express JSON Data router with the provided options.

#### Options

##### data
Type: `Object`

Default: `{}`

The data object upon which the Express JSON Data router will operate.

##### limit
Type: `Number` or `String`

The maximum body size for `PUT`, `POST` and `PATCH` requests.
 
The limit value is passed directly to [bodyParser.json()](https://www.npmjs.com/package/body-parser#limit) without
modification.

##### type
Type: `String` or `Array<String>` or `Function<String, Boolean>`

Default: `[ 'application/json', 'application/json-patch+json' ]`

The supported media types for `PUT`, `POST` and `PATCH` requests.

The type value is passed directly to [bodyParser.json()](https://www.npmjs.com/package/body-parser#type) without
modification.

## HTTP Methods

All of the following HTTP methods may be used with the base URI (e.g. `/data`) or with any path beneath the base URI 
(e.g. `/data/path/to/value`).

If the base URI is used, the HTTP request will operate on the entire data object.
 
If a path beneath the base URI is used, the request path will be treated as a [JSON Pointer](https://tools.ietf.org/html/rfc6901)
to a value within the data object, and the HTTP request will operate on that value only.

### GET

Return the current value of the data object or request path value.

### PUT

Replace the contents of the data object or request path value with the request body.

This method will return the updated value of the data object or request path value.

### POST

Merge the request body into contents of the data object or request path value.

This method uses the [`_.merge`](https://lodash.com/docs#merge) method from [lodash](https://lodash.com/) to merge the 
data, with the request body as the source object. 

This method will return the updated value of the data object or request path value.

### PATCH

Apply the request body as a [JSON Patch](https://tools.ietf.org/html/rfc6902) to the data object or request path value.

This method will accept either an `Array` of JSON patch operations or an `Object` containing a single JSON patch
operation.

This method will return the updated value of the data object or request path value.

### DELETE

Clear the contents of the data object, or remove the request path value from the data object.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [4.1.0](https://github.com/kjots/express-json-data/compare/v4.0.0...v4.1.0) (2021-11-10)


### Features

* update to use Babel 7 ([0f3793d](https://github.com/kjots/express-json-data/commit/0f3793d9d9e74da390429aacf3dde03e3bfbc5a3))

## [4.0.0](https://github.com/kjots/express-json-data/compare/v3.2.1...v4.0.0) (2021-10-31)


### ⚠ BREAKING CHANGES

* the minimum supported Node version is Node 14 LTS.

### Features

* update mimumum supported Node version to be Node 14 LTS ([a1803f5](https://github.com/kjots/express-json-data/commit/a1803f591faec8a1b51b29b43a7a958328be9680))


### Bug Fixes

* update to use fast-json-patch version 3.1.0 ([25be3a7](https://github.com/kjots/express-json-data/commit/25be3a73083e3cb4a66c11bb569234f40c77b98c))


test/init.js
import 'babel-polyfill';

import chai from 'chai';
import chaiAsPromised from 'chai-as-promised';
import sinonChai from 'sinon-chai';

chai.use(chaiAsPromised);
chai.use(sinonChai);

global.expect = chai.expect;

src/index.spec.js
import chai from 'chai';
import chaiAsPromised from 'chai-as-promised';
import nodeMocksHttp from 'node-mocks-http';
import sinon from 'sinon';
import sinonChai from 'sinon-chai';

import expressJsonData from './index.js';

const { expect } = chai
    .use(chaiAsPromised)
    .use(sinonChai);

let sandbox;

beforeEach(() => sandbox = sinon.createSandbox());
afterEach(() => sandbox.restore());

describe('expressJsonData()', () => {
    let data;

    beforeEach(() => {
        data = { testItem: { testItemKey: 'Test Item Value' } };
    });

    it('should create the express router', () => {
        // Given

        // When
        const router = expressJsonData();

        // Then
        expect(router).to.not.be.undefined;
    });

    context('on GET /', () => {
        it('should respond with the data', () => {
            // Given
            let req = nodeMocksHttp.createRequest({ method: 'GET', url: '/' });
            let res = nodeMocksHttp.createResponse();

            sandbox.spy(res, 'json');

            // When
            expressJsonData({ data })(req, res);

            // Then
            expect(res.json).to.have.been.calledWith({ testItem: { testItemKey: 'Test Item Value' } });
        });
    });

    context('on GET /*', () => {
        context('when the data item exists', () => {
            it('should respond with the data item', () => {
                // Given
                let req = nodeMocksHttp.createRequest({ method: 'GET', url: '/testItem' });
                let res = nodeMocksHttp.createResponse();

                sandbox.spy(res, 'json');

                // When
                expressJsonData({ data })(req, res);

                // Then
                expect(res.json).to.have.been.calledWith({ testItemKey: 'Test Item Value' });
            });
        });

        context('when the data item does not exist', () => {
            it('should respond with HTTP 404 Not Found', () => {
                // Given
                let req = nodeMocksHttp.createRequest({ method: 'GET', url: '/notTestItem' });
                let res = nodeMocksHttp.createResponse();

                sandbox.spy(res, 'status');
                sandbox.spy(res, 'end');

                // When
                expressJsonData({ data })(req, res);

                // Then
                expect(res.status).to.have.been.calledWith(404);
                expect(res.end).to.have.been.called;
            });
        });
    });

    context('on DELETE /', () => {
        it('should remove the data', () => {
            // Given
            let req = nodeMocksHttp.createRequest({ method: 'DELETE', url: '/' });
            let res = nodeMocksHttp.createResponse();

            sandbox.spy(res, 'status');
            sandbox.spy(res, 'end');

            // When
            expressJsonData({ data })(req, res);

            // Then
            expect(data).to.eql({});

            expect(res.status).to.have.been.calledWith(204);
            expect(res.end).to.have.been.called;
        });
    });

    context('on DELETE /*', () => {
        context('when the data item exists', () => {
            it('should remove the data item', () => {
                // Given
                let req = nodeMocksHttp.createRequest({ method: 'DELETE', url: '/testItem' });
                let res = nodeMocksHttp.createResponse();

                sandbox.spy(res, 'status');
                sandbox.spy(res, 'end');

                // When
                expressJsonData({ data })(req, res);

                // Then
                expect(data).to.eql({});

                expect(res.status).to.have.been.calledWith(204);
                expect(res.end).to.have.been.called;
            });
        });

        context('when the data item does not exist', () => {
            it('should respond with HTTP 404 Not Found', () => {
                // Given
                let req = nodeMocksHttp.createRequest({ method: 'DELETE', url: '/notTestItem' });
                let res = nodeMocksHttp.createResponse();

                sandbox.spy(res, 'status');
                sandbox.spy(res, 'end');

                // When
                expressJsonData({ data })(req, res);

                // Then
                expect(res.status).to.have.been.calledWith(404);
                expect(res.end).to.have.been.called;
            });
        });
    });

    context('on PUT /', () => {
        it('should replace the data with the request data', () => {
            // Given
            let req = nodeMocksHttp.createRequest({ method: 'PUT', url: '/', body: { newTestItem: { newTestItemKey: 'New Test Item Value' } } });
            let res = nodeMocksHttp.createResponse();

            sandbox.spy(res, 'status');
            sandbox.spy(res, 'json');

            // When
            expressJsonData({ data })(req, res);

            // Then
            expect(data).to.eql({ newTestItem: { newTestItemKey: 'New Test Item Value' } });

            expect(res.status).to.have.been.calledWith(201);
            expect(res.json).to.have.been.calledWith({ newTestItem: { newTestItemKey: 'New Test Item Value' } });
        });
    });

    context('on PUT /*', () => {
        it('should replace the data item with the request data', () => {
            // Given
            let req = nodeMocksHttp.createRequest({ method: 'PUT', url: '/testItem', body: { newTestItemKey: 'New Test Item Value' } });
            let res = nodeMocksHttp.createResponse();

            sandbox.spy(res, 'status');
            sandbox.spy(res, 'json');

            // When
            expressJsonData({ data })(req, res);

            // Then
            expect(data).to.eql({ testItem: { newTestItemKey: 'New Test Item Value' } });

            expect(res.status).to.have.been.calledWith(201);
            expect(res.json).to.have.been.calledWith({ newTestItemKey: 'New Test Item Value' });
        });
    });

    context('on POST /', () => {
        it('should merge the request data into the data', () => {
            // Given
            let req = nodeMocksHttp.createRequest({ method: 'POST', url: '/', body: { newTestItem: { newTestItemKey: 'New Test Item Value' } } });
            let res = nodeMocksHttp.createResponse();

            sandbox.spy(res, 'status');
            sandbox.spy(res, 'json');

            // When
            expressJsonData({ data })(req, res);

            // Then
            expect(data).to.eql({ testItem: { testItemKey: 'Test Item Value' }, newTestItem: { newTestItemKey: 'New Test Item Value' } });

            expect(res.status).to.have.been.calledWith(201);
            expect(res.json).to.have.been.calledWith({ testItem: { testItemKey: 'Test Item Value' }, newTestItem: { newTestItemKey: 'New Test Item Value' } });
        });
    });

    context('on POST /*', () => {
        it('should merge the request data into the data item', () => {
            // Given
            let req = nodeMocksHttp.createRequest({ method: 'POST', url: '/testItem', body: { newTestItemKey: 'New Test Item Value' } });
            let res = nodeMocksHttp.createResponse();

            sandbox.spy(res, 'status');
            sandbox.spy(res, 'json');

            // When
            expressJsonData({ data })(req, res);

            // Then
            expect(data).to.eql({ testItem: { testItemKey: 'Test Item Value', newTestItemKey: 'New Test Item Value' } });

            expect(res.status).to.have.been.calledWith(201);
            expect(res.json).to.have.been.calledWith({ testItemKey: 'Test Item Value', newTestItemKey: 'New Test Item Value' });
        });
    });
    
    context('on PATCH /', () => {
        context('when the request body is an object', () => {
            it('should apply the request body to the data as a JSON Patch', () => {
                // Given
                let req = nodeMocksHttp.createRequest({ method: 'PATCH', url: '/', body: { op: 'replace', path: '/testItem', value: { newTestItemKey: 'New Test Item Value' } } });
                let res = nodeMocksHttp.createResponse();

                sandbox.spy(res, 'status');
                sandbox.spy(res, 'json');

                // When
                expressJsonData({ data })(req, res);

                // Then
                expect(data).to.eql({ testItem: { newTestItemKey: 'New Test Item Value' } });

                expect(res.status).to.have.been.calledWith(201);
                expect(res.json).to.have.been.calledWith({ testItem: { newTestItemKey: 'New Test Item Value' } });
            });
        });

        context('when the request body is an array', () => {
            it('should apply the request body to the data as a JSON Patch', () => {
                // Given
                let req = nodeMocksHttp.createRequest({ method: 'PATCH', url: '/', body: [{ op: 'replace', path: '/testItem', value: { newTestItemKey: 'New Test Item Value' } }] });
                let res = nodeMocksHttp.createResponse();

                sandbox.spy(res, 'status');
                sandbox.spy(res, 'json');

                // When
                expressJsonData({ data })(req, res);

                // Then
                expect(data).to.eql({ testItem: { newTestItemKey: 'New Test Item Value' } });

                expect(res.status).to.have.been.calledWith(201);
                expect(res.json).to.have.been.calledWith({ testItem: { newTestItemKey: 'New Test Item Value' } });
            });
        });
    });

    context('on PATCH /*', () => {
        context('when the data item exists', () => {
            context('when the request body is an object', () => {
                it('should apply the request body to the data item as a JSON Patch', () => {
                    // Given
                    let req = nodeMocksHttp.createRequest({ method: 'PATCH', url: '/testItem', body: { op: 'replace', path: '/testItemKey', value: 'New Test Item Value' } });
                    let res = nodeMocksHttp.createResponse();

                    sandbox.spy(res, 'status');
                    sandbox.spy(res, 'json');

                    // When
                    expressJsonData({ data })(req, res);

                    // Then
                    expect(data).to.eql({ testItem: { testItemKey: 'New Test Item Value' } });

                    expect(res.status).to.have.been.calledWith(201);
                    expect(res.json).to.have.been.calledWith({ testItemKey: 'New Test Item Value' });
                });
            });

            context('when the request body is an array', () => {
                it('should apply the request body to the data item as a JSON Patch', () => {
                    // Given
                    let req = nodeMocksHttp.createRequest({ method: 'PATCH', url: '/testItem', body: [{ op: 'replace', path: '/testItemKey', value: 'New Test Item Value' }] });
                    let res = nodeMocksHttp.createResponse();

                    sandbox.spy(res, 'status');
                    sandbox.spy(res, 'json');

                    // When
                    expressJsonData({ data })(req, res);

                    // Then
                    expect(data).to.eql({ testItem: { testItemKey: 'New Test Item Value' } });

                    expect(res.status).to.have.been.calledWith(201);
                    expect(res.json).to.have.been.calledWith({ testItemKey: 'New Test Item Value' });
                });
            });
        });

        context('when the data item does not exist', () => {
            it('should respond with HTTP 404 Not Found', () => {
                // Given
                let req = nodeMocksHttp.createRequest({ method: 'PATCH', url: '/notTestItem' });
                let res = nodeMocksHttp.createResponse();

                sandbox.spy(res, 'status');
                sandbox.spy(res, 'end');

                // When
                expressJsonData({ data })(req, res);

                // Then
                expect(res.status).to.have.been.calledWith(404);
                expect(res.end).to.have.been.called;
            });
        });
    });
});

test/init.spec.js
import sinon from 'sinon';

beforeEach(() => {
    global.sinon = sinon.createSandbox();
});

afterEach(() => {
    global.sinon.restore();

    delete global.sinon;
});

.eslintrc.js
module.exports = {
    env: {
        es6: true,
        node: true
    },
    extends: 'eslint:recommended',
    parserOptions: {
        ecmaVersion: 2017,
        sourceType: 'module'
    },
    rules: {
        'arrow-spacing': [ 'error', { before: true, after: true } ],
        'brace-style': [ 'error', 'stroustrup', { allowSingleLine: true } ],
        'comma-dangle': [ 'error', 'never' ],
        'eqeqeq': [ 'error', 'always' ],
        'indent': [ 'error', 4, { SwitchCase: 1 } ],
        'keyword-spacing': [ 'error', { before: true, after: true } ],
        'linebreak-style': [ 'error', 'unix' ],
        'no-cond-assign': [ 'error', 'always' ],
        'no-empty': [ 'error', { allowEmptyCatch: true } ],
        'no-unneeded-ternary': [ 'error', { defaultAssignment: false } ],
        'no-var': 'error',
        'prefer-template': 'error',
        'quotes': [ 'error', 'single' ],
        'semi': [ 'error', 'always' ],
        'space-before-blocks': [ 'error', 'always' ],
        'strict': [ 'error', 'global' ],
        'yoda': [ 'error', 'never' ]
    },
    overrides: [
        {
            files: [ '**/*.spec.js' ],
            env: {
                mocha: true
            }
        }
    ]
};

gulpfile.babel.js
import 'babel-polyfill';

import chalk from 'chalk';
import chokidar from 'chokidar';
import del from 'del';
import gulp from 'gulp';
import gulpBabel from 'gulp-babel';
import gulpEslint from 'gulp-eslint';
import gulpMocha from 'gulp-mocha';
import gulpSourcemaps from 'gulp-sourcemaps';
import gulpUtil from 'gulp-util';
import path from 'path';

const srcBaseDir = path.resolve('src');
const srcPattern = path.resolve(srcBaseDir, '**/!(*.spec).js');
const specPattern = path.resolve(srcBaseDir, '**/*.spec.js');

const libBaseDir = path.resolve('lib');

function compile(pattern, logErrors) {
    function handleCompileError(error) {
        if (!logErrors) throw error;

        gulpUtil.log(error.toString());

        this.end();
    }

    gulpUtil.log('Compiling', chalk.magenta(pattern));

    return gulp.src(pattern, { base: srcBaseDir })
        .pipe(gulpSourcemaps.init())
        .pipe(gulpBabel()).on('error', handleCompileError)
        .pipe(gulpSourcemaps.write('.', {
            includeContent: false,
            sourceRoot: file => {
                let srcDir = path.dirname(file.path);
                let libDir = path.resolve(libBaseDir, path.relative(srcBaseDir, srcDir));

                return path.relative(libDir, srcBaseDir);
            }
        }))
        .pipe(gulp.dest('lib'));
}

gulp.task('test', [ 'test:eslint', 'test:mocha' ]);

gulp.task('test:eslint', [ 'test:eslint:gulpfile', 'test:eslint:src', 'test:eslint:spec' ]);

gulp.task('test:eslint:gulpfile', () => {
    return gulp.src(__filename)
        .pipe(gulpEslint())
        .pipe(gulpEslint.format())
        .pipe(gulpEslint.failAfterError());
});

gulp.task('test:eslint:src', () => {
    return gulp.src(srcPattern)
        .pipe(gulpEslint())
        .pipe(gulpEslint.format())
        .pipe(gulpEslint.failAfterError());
});

gulp.task('test:eslint:spec', () => {
    return gulp.src(specPattern)
        .pipe(gulpEslint())
        .pipe(gulpEslint.format())
        .pipe(gulpEslint.failAfterError());
});

gulp.task('test:mocha', [], () => {
    return gulp.src(specPattern, { read: false })
        .pipe(gulpMocha({ require: [ 'babel-core/register', 'babel-polyfill' ] }));
});

gulp.task('compile', () => compile(srcPattern));

gulp.task('watch:compile', [ 'compile' ], () => {
    chokidar.watch(srcPattern, { ignoreInitial: true })
        .on('add', srcPath => compile(srcPath, true))
        .on('change', srcPath => compile(srcPath, true))
        .on('unlink', srcPath => {
            let libPath = path.resolve(libBaseDir, path.relative(srcBaseDir, srcPath));

            del([libPath, `${libPath}.map`]);
        });
});

gulp.task('clean', [ 'clean:coverage', 'clean:lib' ]);

gulp.task('clean:coverage', () => del([ '.nyc_output', 'coverage' ]));

gulp.task('clean:lib', () => del([ libBaseDir ]));

src/index.js
import _ from 'lodash';
import bodyParser from 'body-parser';
import express from 'express';
import fastJsonPatch from 'fast-json-patch';

export default function expressJsonData({ data = {}, limit, type = [ 'application/json', 'application/json-patch+json' ] } = {}) {
    function clearData() {
        Object.keys(data).forEach(key => delete data[key]);
    }

    function path2pointer(path) {
        return path
            .split('/')
            .map(component => fastJsonPatch.escapePathComponent(decodeURIComponent(component)))
            .join('/');
    }

    function jsonPointerDecoder(req, res, next) {
        req.pointer = path2pointer(req.path);
        next();
    }

    return express.Router()
        .use(jsonPointerDecoder)
        .get('/', (req, res) => {
            return res.json(data);
        })
        .get('/*', (req, res) => {
            let value = fastJsonPatch.getValueByPointer(data, req.pointer);
            if (value === undefined) {
                return res.status(404).end();
            }

            return res.json(value);
        })
        .delete('/', (req, res) => {
            clearData();

            return res.status(204).end();
        })
        .delete('/*', (req, res) => {
            if (fastJsonPatch.getValueByPointer(data, req.pointer) === undefined) {
                return res.status(404).end();
            }

            fastJsonPatch.applyOperation(data, { op: 'remove', path: req.pointer });

            return res.status(204).end();
        })
        .use(bodyParser.json({ limit, type }))
        .put('/', (req, res) => {
            clearData();

            _.merge(data, req.body);

            return res.status(201).json(data);
        })
        .put('/*', (req, res) => {
            fastJsonPatch.applyOperation(data, { op: 'replace', path: req.pointer, value: req.body });

            return res.status(201).json(req.body);
        })
        .post('/', (req, res) => {
            _.merge(data, req.body);

            return res.status(201).json(data);
        })
        .post('/*', (req, res) => {
            let value = _.merge({}, fastJsonPatch.getValueByPointer(data, req.pointer), req.body);

            fastJsonPatch.applyOperation(data, { op: 'replace', path: req.pointer, value });

            return res.status(201).json(value);
        })
        .patch('/', (req, res) => {
            if (Array.isArray(req.body)) {
                fastJsonPatch.applyPatch(data, req.body);
            }
            else {
                fastJsonPatch.applyOperation(data, req.body);
            }

            return res.status(201).json(data);
        })
        .patch('/*', (req, res) => {
            let value = fastJsonPatch.getValueByPointer(data, req.pointer);
            if (value === undefined) {
                return res.status(404).end();
            }

            if (Array.isArray(req.body)) {
                fastJsonPatch.applyPatch(value, req.body);
            }
            else {
                fastJsonPatch.applyOperation(value, req.body);
            }

            return res.status(201).json(value);
        });
}